cairo: add new version and update build system

Chrismarsh opened the pull request:
Post `cairo@1.17.7:`, cairo requires the meson build system. As a result, the `@1.18` version doesn't build.

This is a pretty significant update to 
- add `@1.18:`
- add meson build system
- unify the new meson variants with existing variant names
- remove `"1.16.0",` from being a preferred version, as there is no reason why it needs to be. It also doesn't build on macos.

There are clearly some assumptions about what meson variants are enabled together, so this tries to protect against incompatible variant options I found in testing.

I couldn't get the new `symbol-lookup` variant to work on macos. Maybe this is a linux specific thing? It is currently `False` by default and conflicts on `+symbol-lookup` such that this can be fixed in the future.

`svg` and `pdf` are now legacy variants. However, many packages still use these variants. The problem is that the new variants `png` and `zlib` combine `+png = +png+svg` and `+zlib=ps, pdf, ...`. I have made the decision that `+svg and +pdf` can turn on the meta variants, but cannot disable them.

Chrismarsh committed:
update cairo for new meson build system

Chrismarsh committed:
update patch range. remove old conflict

Chrismarsh commented:
@spackbot fix style

Chrismarsh committed:
style

wdconinc commented:
See also https://github.com/spack/spack/pull/43586, in particular some failure modes observed by @AlexanderRichert-NOAA .

Chrismarsh committed:
update pango to reflect the changes in cairo

Chrismarsh committed:
refine depends

Chrismarsh committed:
style

Chrismarsh committed:
add lzo depends

Chrismarsh commented:
@wdconinc thanks for pointing that PR out, not sure how I missed it. 

To @AlexanderRichert-NOAA's [issue](https://github.com/spack/spack/pull/43586#pullrequestreview-1998129640), _without_ the extra meson cflag addition, I can build
```
spack install cairo@1.18.2 +ft ^freetype~shared
```
on macos and linux. However, there are a bunch of, seemingly (?) assumptions in the meson build such that `+fc +ft +zlib +png` should all be enabled together to get a clean build -- it is what I set as the default behaviour.

I have also added a fix to `pango` to this commit to ensure the `+quartz` variant is used from this new cairo on darwin.

Chrismarsh committed:
add +shared

Chrismarsh committed:
non self-referential variant requireme

Chrismarsh committed:
style

Chrismarsh committed:
Move +shared variant back to just autotools as meson automatically handles it

tgamblin commented on var/spack/repos/builtin/packages/cairo/package.py:
```diff
@@ -25,55 +27,187 @@ class Cairo(AutotoolsPackage):
     patch("fontconfig.patch", when="@1.16.0:1.17.2")
     # Don't regenerate docs to avoid a dependency on gtk-doc
-    patch("disable-gtk-docs.patch", when="^autoconf@2.70:")
+    patch("disable-gtk-docs.patch", when="@:1.17.4^autoconf@2.70:")
```
I think this might be clearer:

```suggestion
    patch("disable-gtk-docs.patch", when="build_system=meson ^autoconf@2.70:")
```

tgamblin replied:
Also while it's not possible right now, when we switch to compilers as nodes, I think this can more clearly become:

```python
    patch("disable-gtk-docs.patch", when="%autoconf@2.70:")
```

`^` (transitive link/run OR direct build dep)  will also work properly but the new semantics for `%` (direct build dep) is clearer.   Right now `^` is "any transitive dep" which is likely why this wasn't constrained enough.

Chrismarsh replied:
Yes I like that. n.b., that should be `build_system=autotools`. I have made the change

Chrismarsh committed:
clarify patch when=

wdconinc commented on var/spack/repos/builtin/packages/cairo/package.py:
```diff
@@ -25,55 +27,187 @@ class Cairo(AutotoolsPackage):
+
 
+class AutotoolsBuilder(autotools.AutotoolsBuilder):
     def autoreconf(self, spec, prefix):
```
When in AutotoolsBuilder, this requires pkg as argument.
```suggestion
    def autoreconf(self, pkg, spec, prefix):
```

Chrismarsh replied:
Fixed thanks

wdconinc commented on var/spack/repos/builtin/packages/cairo/package.py:
```diff
@@ -25,55 +27,187 @@ class Cairo(AutotoolsPackage):
+
+        # +pdf implies zlib now
+        if self.spec.satisfies("+pdf"):
+            self.spec.variants["zlib"].value = True
```
This seems like it could get confusion when stuff is changed under the feet of the users. Can we add a conflict for this instead?
```py
    conficts("~zlib", when="+pdf build_system=meson")
```

Chrismarsh replied:
resolved

Chrismarsh replied:
I've replaced these with requires and refined the conflicts. I think given the sane defaults, the conflicts and requires should keep it mostly in line.

wdconinc commented on var/spack/repos/builtin/packages/cairo/package.py:
```diff
@@ -25,55 +27,187 @@ class Cairo(AutotoolsPackage):
+
+        # +svg implies png now
+        if self.spec.satisfies("+svg"):
+            self.spec.variants["png"].value = True
```
Same here as with pdf zlib:
```py
    conflicts("~png", when="+svg build_system=meson")
```

Chrismarsh replied:
Fixed

Chrismarsh committed:
update based on reviews. switch from conflicts to requires to enforce variant synchronization

Chrismarsh committed:
refine conflicts and requires

wdconinc commented on var/spack/repos/builtin/packages/cairo/package.py:
```diff
@@ -25,56 +27,198 @@ class Cairo(AutotoolsPackage):
+    # https://github.com/microsoft/vcpkg/pull/38313
+    depends_on("lzo", when="@1.17.6: build_system=meson")
+
+    with when("@:1.17.6"):
```
```suggestion
    with when("build_system=autotools"):
```

Chrismarsh replied:
good idea. I've moved all the variant and build under one `when("build_system=autotools"):` and then kept the version specific depends on their own `with when`

wdconinc replied:
I am still confused by this (and similar below). You now have:
```
    # versions that use (the old) autotools build
    with when("@:1.17.6"):
```
but if this is supposed to apply to autotools builds (per comment), then we should use `with when` on the `build_system` variant here. It is even more confusing because you have a different version here (autotools is valid when `@:1.17.4`, not `@:1.17.6`), and it is incorrect since the meson build requires e.g. `glib@2.14:` starting in 1.17.4 per [upstream](https://gitlab.freedesktop.org/cairo/cairo/-/commit/596a82f2d185b101bd74645492821fe2f9e0daa0#0cc1139e3347f573ae1feee5b73dbc8a8a21fcfa).

Right now, you could concretize `cairo@=1.17.6 build_system=meson ^glib@2.13` and it would fail on compilation because the glib is too old, and same with libpng, freetype, etc.

Chrismarsh replied:
I was under the impression that the `build_system(...)` selection pined the build system to specific version ranges. This is clearly not the case! 

I have added a `requires` that ensures the build system is enforced to the correct version ranges.

I had been using [1.17.6](https://gitlab.freedesktop.org/cairo/cairo/-/tree/1.17.6?ref_type=tags) as it has both meson and autotools builds. However, it isn't in spack, it isn't tested in spack, and just overcomplicates matters. I have moved all autotools references to `@:1.17.4` and increased the meson references to `@1.18.0:`.

The way I have this laid out with the `with when(...)` is:
use ` with when("build_system=XXX"):` for the build-system specific variants and build depends


and then 
`with when("@:1.XX"):` for the version specific, non-build depends.

My intention was to make it easier to add new non-build `depends_on` as new versions of cairo were added. I've added some comments  to further clarify.

This may be a "makes sense to me" but not a spackonic way of doing things -- I can combine the `with whens` if you prefer.

wdconinc commented on var/spack/repos/builtin/packages/cairo/package.py:
```diff
@@ -25,56 +27,198 @@ class Cairo(AutotoolsPackage):
+        depends_on("glib")
+        depends_on("which", type="build")
+
+    with when("@1.17.8:"):
```
```suggestion
    with when("build_system=meson"):
```

wdconinc reviewed (dismissed):
Two more minor comments to avoid repetition of versions where it's the biuld system that is relevant.

Chrismarsh committed:
better group build deps together

wdconinc commented:
1. Can you comment on whether the changing variants make any package dependencies within the builtin repository effectively pinned at the autotools version? 
2. There is now autotools until 1.17.4 and meson as of 1.17.6. But in L149, you have changing dependencies from before 1.17.6 and after 1.17.8. Are these cutoff versions between old and new behavior intended to be the same?

Chrismarsh committed:
comment for meson build lower version bound

Chrismarsh committed:
clarifying comments

Chrismarsh commented:
1. >Can you comment on whether the changing variants make any package dependencies within the builtin repository effectively pinned at the autotools version?

There is no pinned autotools version (from what I can see or ^f):
https://github.com/spack/spack/pull/48822/files#diff-c711ae2c1dcb9cfaef60e6dfe49614cd68924ef0fdc9e77bcb1f438df3bacaaeR40

https://github.com/spack/spack/pull/48822/files#diff-c711ae2c1dcb9cfaef60e6dfe49614cd68924ef0fdc9e77bcb1f438df3bacaaeR65

Am I missing something?



2. >There is now autotools until 1.17.4 and meson as of 1.17.6. But in L149, you have changing dependencies from before 1.17.6 and after 1.17.8. Are these cutoff versions between old and new behavior intended to be the same?

In the lines i changed, e.g., 
```
    with when("@:1.17.6"):
        depends_on("automake", type="build")
        depends_on("autoconf", type="build")
        depends_on("libtool", type="build")
```
`1.17.6` is the `cairo` version, not the `autotools` version, nor the meson version. Previously the cairo version was used to select the build system depends, instead of the `buildsystem` directly. 

I changed it to group build system tooling to each buildsystem , as I agree it is more clear.

So there is no autotools version pin, and [`meson@1.3.0:`](https://github.com/spack/spack/pull/48822/files#diff-c711ae2c1dcb9cfaef60e6dfe49614cd68924ef0fdc9e77bcb1f438df3bacaaeR141) is from here 
https://gitlab.freedesktop.org/cairo/cairo/-/blob/1.18.2/meson.build?ref_type=tags#L2

I've added more comments to hopefully make this more clear, please let me know what you think

Chrismarsh committed:
clarify version ranges, enforce build_system with version ranges

Chrismarsh committed:
style

wdconinc commented on var/spack/repos/builtin/packages/cairo/package.py:
```diff
@@ -25,56 +27,208 @@ class Cairo(AutotoolsPackage):
+        default="meson",
+    )
+
+    requires("build_system=meson", when="@1.18.0:")
+    requires("build_system=autotools", when="@:1.17.4")
+
```
This is not needed since it is the only solution.
```suggestion
```

wdconinc committed:
cairo:  no need to require for build_systems
